### DISCUSSION TITLE
Is it possible to do two shard maps at the same time?

### DISCUSSION THREAD

**Discussion by jeffgortmaker** [jax-ml/jax#23015] [Q&A]
I have data indexed by two dimensions, and I'm trying to shard computation along both of them at the same time. My understanding is that the current recommendation is to use shard maps, but I'm struggling with how to do use them along two dimensions.

Here's a minimal working example of what I'm trying to do. My goal in function `h` is to map a second function `g` over the first dimension `a` of a 2x2 array. Within `g`, I attempt to map a third function `f` over the array's second dimension `b`. Each map applies `shard_map` to `lax.map`:

```python
import os
import jax

print(jax.__version__)  # 0.4.31
jax.config.update('jax_platform_name', 'cpu')
os.environ['XLA_FLAGS'] = '--xla_force_host_platform_device_count=4'

from functools import partial
from jax.sharding import Mesh, PartitionSpec as P
from jax.experimental.shard_map import shard_map
from jax.experimental.mesh_utils import create_device_mesh

mesh = Mesh(create_device_mesh((2, 2)), ('a', 'b'))

def f(x):
    return x + 1

def g(x):
    return shard_map(partial(jax.lax.map, f), mesh, P('b'), P('b'))(x) + 2

def h(x):
    return shard_map(partial(jax.lax.map, g), mesh, P('a'), P('a'))(x) + 3

import numpy as np
print(g(np.zeros(2)))  # [3. 3.]
print(h(np.zeros((2, 2))))  # error
```

With the current JAX version of 0.4.31, this raises: 

    AttributeError: 'RewriteTrace' object has no attribute 'process_shard_map'

If I set `check_rep=False` in each `shard_map`, it instead raises:

    ValueError: Axis: b of PartitionSpec('b',) is also found in manual_axes: frozenset({'b', 'a'})

If I replace `h` with `jax.lax.map(g, x) + 3`, I get a 2x2 array of sixes as expected. But computation is only sharded over the second axis.

My guess is there's something terribly wrong with how I'm trying to do this. Any advice would be wonderful!

**Comment by jakevdp** [comment]
xref #21004, which is a similar question.

Maybe @yashk2810 has some ideas here?

**Reply by yashk2810** [reply]
Yeah, you need to use `auto` parameter. For example, this works:

```
In [6]: from functools import partial
   ...: from jax.sharding import Mesh, PartitionSpec as P
   ...: from jax.experimental.shard_map import shard_map
   ...: from jax.experimental.mesh_utils import create_device_mesh
   ...:
   ...: mesh = Mesh(create_device_mesh((2, 2)), ('a', 'b'))
   ...:
   ...: def f(x):
   ...:     return x + 1
   ...:
   ...: def g(x):
   ...:     return shard_map(f, mesh, P('b'), P('b'))(x) + 2
   ...:
   ...: def h(x):
   ...:     return shard_map(g, mesh, P('a'), P('a'), check_rep=False, auto=frozenset({'b'}))(x) + 3
   ...:
   ...: import numpy as np
   ...: print(g(np.zeros(4)))
   ...: print(jax.jit(h)(np.zeros((4))))
[3. 3. 3. 3.]
[6. 6. 6. 6.]
```

**Reply by jeffgortmaker** [reply]
This seems really promising! I'm struggling to apply it to the original example, though. When I simply add `auto=frozenset({'b'})` after setting `check_rep=False`, it raises a `NotImplementedError` from the start of `jax.experimental.shard_map._shard_map_impl`, and when I wrap `h` in a `jit` like in your example, I get a sort of confusing error:

    F external/xla/xla/hlo/utils/hlo_sharding_util.cc:2806] Check failed: sharding.IsManualSubgroup() 

Thoughts on how to properly use the `auto` parameter in the original example?

**Reply by yashk2810** [reply]
yeah, I saw that too. Might be some interaction with `lax.map`. I can send a repro to @pschuh

**Reply by yashk2810** [reply]
have you tried using `vmap` instead? What's the reason for using `lax.map`?

**Reply by jeffgortmaker** [reply]
Thanks for sending the repro! Hoping we can figure it out.

I'm using `lax.map` because the analogue of the `f` function in my actual code doesn't vectorize well. It includes a `while` loop that can take anywhere from one to a thousand iterations; the `vmap` batching rule (I think that's what it's called?) for this then ends up always doing a thousand iterations for each call, while `lax.map` allows it to only do as many iterations as are necessary.

If you're interested, I've found this setting to show up often in simulation estimators for economic models. Dimension `a` is over sets of simulation draws, dimension `b` is over economic agents, and `f` solves each agent's optimization problem, which can be easy for some agents and harder for others.

**Reply by yashk2810** [reply]
Makes sense. If possible, can you file this on github issue tracker? or somehow convert this discussion into an issue?

**Reply by jeffgortmaker** [reply]
I just [learned](https://github.com/orgs/community/discussions/2861) that GitHub doesn't support the discussion $\to$ issue direction, but I filed a new https://github.com/google/jax/issues/23019.
